harness/determined/core/_checkpoint.py

```diff
@@ -34,6 +34,107 @@ class DownloadMode(enum.Enum):
     NoSharedDownload = "NO_SHARED_DOWNLOAD"
 
 
+def merge_metadata(all_metadata: List[Dict[str, Any]]) -> Tuple[Dict[str, Any], Dict[str, List]]:
+    """
+      Given a list of metadata, return:
+        - merged metadata
+        - a dict mapping conflicting keys to ranks that would upload them
+
+    Merging scenarios:
+    #  Metadata 1        | Metadata 2          | Conflict |  Result
+    # -------------------|---------------------|----------|-------------------------------
+    #  key1: val         | key2: val           | no       | {key1: val, key2: val}
+    #  key1: val         | key1: val           | yes      | n/a
+    #  key1: [val]       | key1: [val]         | no       | {key1: [val, val]}
```

Comment:
I think lists should not be automatically merged.

I think there is actually a [published RFC about json merging](https://www.rfc-editor.org/rfc/rfc7386.html), and they don't allow merging lists, and I think we should stick to that.

Users who want to merge an array of objects from each rank should emit this structure from each RANK_ID:

```
{"$RANK_ID": my_item}
```

That way the user isn't allowed to do anything that isn't really obvious how it will merge.

If for instance, each rank emitted a varying length list, then you would have no idea about the bounds of the each rank's contribution after merging.

Reply:
> If for instance, each rank emitted a varying length list, then you would have no idea about the bounds of the each rank's contribution after merging.

fair point!  yeah, in RFC list are not getting any special treatment when merging. Instead, they are overwritten with new values. In our case, that means a conflict, which I am okay with.

Let's merge dictionaries only and treat lists and values the same.